File: src/main/java/com/company/erp/features/chats/dto/MessageDto.java
package com.company.erp.features.chats.dto;

import com.company.erp.features.chats.entity.Message;
import com.company.erp.features.chats.entity.MessageType;

import java.time.Instant;
import java.util.List;
import java.util.Set;

public record MessageDto(
        Long id,
        Long conversationId,
        Long senderId,
        MessageType type,
        String body,
        String attachmentUrl,
        String attachmentContentType,
        Long attachmentSizeBytes,
        Integer durationSeconds,
        Long replyToMessageId,
        Instant editedAt,
        boolean deleted,
        List<ReactionDto> reactions,
        /**
         * User ids of conversation members (other than the sender) whose
         * {@code lastReadMessageId} is {@code >=} this message's id.
         * Derived at read time — no separate reads table.
         */
        Set<Long> readByUserIds,
        Instant createdAt
) {
    public static MessageDto from(Message m, List<ReactionDto> reactions) {
        return from(m, reactions, Set.of());
    }

    public static MessageDto from(Message m, List<ReactionDto> reactions, Set<Long> readByUserIds) {
        return new MessageDto(
                m.getId(),
                m.getConversationId(),
                m.getSenderId(),
                m.getType(),
                m.isDeleted() ? null : m.getBody(),
                m.isDeleted() ? null : m.getAttachmentUrl(),
                m.isDeleted() ? null : m.getAttachmentContentType(),
                m.isDeleted() ? null : m.getAttachmentSizeBytes(),
                m.isDeleted() ? null : m.getDurationSeconds(),
                m.getReplyToMessageId(),
                m.getEditedAt(),
                m.isDeleted(),
                reactions,
                readByUserIds,
                m.getCreatedAt());
    }
}

File: app/Features/Chat/Dto/MessageDto.php
<?php

namespace App\Features\Chat\Dto;

use App\Features\Chat\Models\ChatMessage;

/**
 * Port of the Spring MessageDto record.
 *
 * When the message is soft-deleted, body and attachment fields are nulled out
 * (matching the Java DTO). readByUserIds carries the user ids of conversation
 * members (other than the sender) whose lastReadMessageId is >= this message's
 * id — derived at read time, no separate reads table.
 */
final class MessageDto
{
    /**
     * @param array<int, array<string, mixed>> $reactions
     * @param array<int, int> $readByUserIds
     * @return array<string, mixed>
     */
    public static function from(ChatMessage $m, array $reactions = [], array $readByUserIds = []): array
    {
        $deleted = $m->isDeleted();

        return [
            'id' => (int) $m->id,
            'conversationId' => (int) $m->conversation_id,
            'senderId' => $m->sender_id !== null ? (int) $m->sender_id : null,
            'type' => $m->type,
            'body' => $deleted ? null : $m->body,
            'attachmentUrl' => $deleted ? null : $m->attachment_url,
            'attachmentContentType' => $deleted ? null : $m->attachment_content_type,
            'attachmentSizeBytes' => $deleted ? null : ($m->attachment_size_bytes !== null ? (int) $m->attachment_size_bytes : null),
            'durationSeconds' => $deleted ? null : ($m->duration_seconds !== null ? (int) $m->duration_seconds : null),
            'replyToMessageId' => $m->reply_to_message_id !== null ? (int) $m->reply_to_message_id : null,
            'editedAt' => $m->edited_at ? $m->edited_at->utc()->format('Y-m-d\TH:i:s\Z') : null,
            'deleted' => $deleted,
            'reactions' => array_values($reactions),
            'readByUserIds' => array_values(array_map('intval', $readByUserIds)),
            'createdAt' => $m->created_at ? $m->created_at->utc()->format('Y-m-d\TH:i:s\Z') : null,
        ];
    }
}
